File: 백준/Silver/2805. 나무 자르기/나무 자르기.cc
#include <iostream>

using namespace std;
using int64 = long long;

int N, M;
int logs[1000002];

bool solve(int64 mid)
{
    int64 ans = 0;
    for (int i = 0; i < N; ++i) {
        if (logs[i] <= mid)
            continue;

        ans += logs[i] - mid;
        if (ans >= M)
            return true;
    }
    return false;
}

int main(void)
{
    cin.tie(0)->tie(0)->sync_with_stdio(0);
    
    cin >> N >> M;
    for (int i = 0; i < N; ++i)
        cin >> logs[i];

    int64 st = 0, en = 1e9;
    int ans = 0;

    while (st <= en) {
        int64 mid = (st + en) / 2;

        if (solve(mid)) {
            st = mid + 1;
            ans = mid;
        }
        else {
            en = mid - 1;
        }
    }

    cout << en;
}

File: 백준/Silver/2805. 나무 자르기/나무 자르기.js
/**
 * Input
 */
const filePath = process.platform === "linux" ? "/dev/stdin" : "./src/input.txt";
const input = require("fs").readFileSync(filePath).toString().trim().split("\n");

const [N, M] = input[0].split(" ").map(Number);
const arr = input[1].split(" ").map(Number);

/**
 * Solution
 */

let st = 1, end = 1000000000;

function solve(len) {
    let cut = 0;
    arr.forEach(e => {
        if (e >= len)
            cut += e - len;
    });
    return cut >= M;
}

let ans = 0
while (st <= end) {
    const mid = Math.trunc((st + end + 1) / 2);

    if (solve(mid)) {
        ans = mid;
        st = mid + 1;
    }
    else
        end = mid - 1;
}

console.log(ans);
